hpcgap/lib/oper.g

```diff
@@ -1276,6 +1276,106 @@ BIND_GLOBAL( "DeclareAttribute", function ( arg )
+##  GAP mutable attributes safe to use in a multi-threaded way, without
+##  significant code changes. If possible, no new thread local attributes should
+##  be added, and existing ones removed and cleaned up where possible.
+##
+##  Current Limitations:
+##
+##  In HPCGAP we disable the ImagesSet overload in lib/ghomperm.gi as it filters
+##  on HasStabChainMutable
+##
+##
+##  Thread local attributes must be mutable. The mutflag option is kept
+##  to keep function calls for DeclareAttribute and DeclareThreadLocalAttribute
+##  as close as possible
+##
+##  This function takes <name> and <filter> and <mutflag>, like DeclareAttribute,
+##  and creates functions called <name>, Set<name> and Has<name>, 
+##  like normal Attributes.
+##
+##  The main difference from normal attributes is that rather than calling
+##  InstallMethod on <name>, InstallMethod must be called on HPCGAP_TL<name>.
+##
+##  In non-HPC-GAP, DeclareThreadLocalAttribute behaves like
+##  DeclareAttribute, except it binds HPCGAP_TL<name> to <name>.
+##  This means InstallMethod can be called as normal, on both
+##  HPCGAP_TL<name> and <name>.
+
+if IsBound(HPCGAP) then
+    BIND_GLOBAL("DeclareThreadLocalAttribute", 
+        function(name, filter, mutflag)
+            local nname, checkList, findmethod;
+
+            if mutflag <> "mutable" then
+                ErrorNoReturn("Thread local attributes must be mutable");
+            fi;
+
+            name := IMMUTABLE_COPY_OBJ(name);
+            checkList := function(obj)
+                if not IsBound(obj!.(name)) then
+                    obj!.(name) := AtomicList(1);
+                fi;
+            end;
+
+            nname := "HPCGAP_TL"; APPEND_LIST_INTR( nname, name );
+            DeclareOperation( nname, [ filter ] );
+            findmethod := VAL_GVAR(nname);
+
+            BIND_GLOBAL(name,
+                function(obj)
+                    local val, threadid;
+                    threadid := ThreadID(CurrentThread())+1;
+                    checkList(obj);
+                    if IsBound(obj!.(name)[threadid]) then
+                    return obj!.(name)[threadid];
+                    fi;
+                    val := findmethod(obj);
+                    obj!.(name)[threadid] := val;
+                    return val;
+                end);
+
+            nname:= "Set"; APPEND_LIST_INTR( nname, name );
+            BIND_GLOBAL( nname,
+                function(obj, val)
+                    checkList(obj);
+                    obj!.(name)[ThreadID(CurrentThread())+1] := val;
+                end);
+
+            nname:= "Has"; APPEND_LIST_INTR( nname, name );
+            BIND_GLOBAL( nname,
+                function(obj)
+                    checkList(obj);
+                    return IsBound(obj!.(name)[ThreadID(CurrentThread())+1]);
+                end);
+        end);
+else
+    BIND_GLOBAL("DeclareThreadLocalAttribute",
+        function(name, filter, mutflag)
+            local nname;
```

Comment:
I'd also insert the `ErrorNoReturn("Thread local attributes must be mutable");` here